File: cocos2d-x-2.2.0/projects/Ironman/Classes/GameScenePlayLayer.cpp
//
//  GameScenePlayLayer.cpp
//  Ironman
//
//  Created by Ken on 13-10-10.
//
//

#include "GameScenePlayLayer.h"
#include "GameScene.h"
#include "Laser.h"
#include "AudioPlayer.h"
#define ANIME_RUN 0
#define ANIME_JUMP 0

bool GameScenePlayLayer::init()
{
	playerX = 50.0;
    playerY = 70.0;
	playerScale = 0.6f;
	touchTime = 0;
	isAttack = false;
	imManArmatureBrood = 100;
	
	monsterGroundAmount = 0;
    monsterSkyAmount = 0;
    
    _attackPos = CCPoint(ccp(0,0));
    _attackDir = 0.0f;

	this->IMRunningStop();
	this->setTouchEnabled(true);
	actionNum = ACTION_RUN_STOP;
    
    return true;
}

void GameScenePlayLayer::registerWithTouchDispatcher(void)
{
    CCDirector::sharedDirector()->getTouchDispatcher()->addTargetedDelegate(this, 0, false);
}

void GameScenePlayLayer::runJumpActionCallBack(CCNode* sender, void* data)
{
	imManArmature->stopAllActions();
    imManArmature->removeFromParentAndCleanup(false);
	if(0xbebabebb ==(unsigned int) data)
	{
		 this ->IMRunning();
	}
	else
	{
		this ->IMRunningStop();
	}
}

void GameScenePlayLayer::standJumpActionCallBack(CCNode* sender, void* data)
{
	imManArmature->stopAllActions();
    imManArmature->removeFromParentAndCleanup(false);
    this ->IMRunningStop();
}

void GameScenePlayLayer::menuCloseCallback(CCObject* pSender)
{
    CCActionInterval * splitCols = CCMoveTo::create(1.0,CCPointMake(imManArmature->getPosition().x+300, imManArmature->getPosition().y));
    imManArmature->runAction(splitCols);
}

bool GameScenePlayLayer::ccTouchBegan(CCTouch *pTouch, CCEvent *pEvent)
{
    m_tBeginPos = pTouch->getLocation();
	return true;
}

void GameScenePlayLayer::ccTouchMoved(CCTouch *pTouch, CCEvent *pEvent)
{
	touchTime++;	
}

void GameScenePlayLayer::ccTouchEnded(CCTouch *pTouch, CCEvent *pEvent)
{
	if(isAttack)
	{
		return;
	}

	if(touchTime>30)
	{
		touchTime = 0;
		return;
	}

    CCPoint touchLocation  = pTouch->getLocation();
	CCPoint setBtnLocation = GameScene::shareGameScene()->menuLayer->settingBtn->getPosition();
	CCSize  setBtnSize = GameScene::shareGameScene()->menuLayer->settingBtn->getSize();
	
	if(touchTime<8 && checkIfTouchNotInSetBtnArea(touchLocation,setBtnSize, setBtnLocation))
	{
		if(actionNum != ACTION_RUN && actionNum != ACTION_RUN_STOP)
		{
			return;
		}
		isAttack = true;
		imManArmature->stopAllActions();
		imManArmature->removeFromParentAndCleanup(false);
		if(actionNum == ACTION_RUN)
		{
			this ->IMRunAttack(touchLocation);
		}
		else if(actionNum == ACTION_RUN_STOP)
		{
			this ->IMStandAttack(touchLocation);
		}
		
		touchTime = 0;
		return;
	}

    float nMoveX = touchLocation.x - m_tBeginPos.x;
	float nMoveY = touchLocation.y - m_tBeginPos.y;
	int radian = 10;
	touchTime = 0;

	if(nMoveX>10 && fabs(tan(nMoveY/nMoveX))<fabs(sqrt(3)/radian))
	{
		if(actionNum == ACTION_RUN)
		{
			if(GameSceneMapLayer::g_map_move_speed!=6)
			{
				GameScene::shareGameScene()->gameSceneMapLayer->setMovedSpeed(6);
				imManArmature->getAnimation()->setSpeedScale(3.0f);
				this->schedule(schedule_selector(GameScenePlayLayer::changeSpeed),0.5f, 0, 0.0f);
				return;
			}
		}
		imManArmature->stopAllActions();
		imManArmature->removeFromParentAndCleanup(false);
		this ->IMRunning();
	}
    
	if(nMoveX<-10 && fabs(tan(nMoveY/nMoveX))<fabs(sqrt(3)/radian))
	{
		if(actionNum == ACTION_RUN_STOP)
			return;
		imManArmature->stopAllActions();
		imManArmature->removeFromParentAndCleanup(false);
		this ->IMRunningStop();
	}
    
	if(nMoveY>10 && fabs(tan(nMoveY/nMoveX))>fabs(sqrt(3)/radian))
	{
		if(actionNum == ACTION_STAND_JUMP || actionNum == ACTION_RUN_JUMP)
			return;
        
        std::string armatureName = imManArmature->getName();
		imManArmature->stopAllActions();
		imManArmature->removeFromParentAndCleanup(false);
        
        if(0 == armatureName.compare("IMRun"))
        {
            this->IMRunJump();
            CCActionInterval * jumpAction = CCJumpTo::create(0.5,CCPointMake(imManArmature->getPosition().x,imManArmature->getPosition().y),200,1);
            CCCallFunc * callBack;
			if(nMoveX<0)
			{
				callBack = CCCallFuncND::create(this, callfuncND_selector(GameScenePlayLayer::runJumpActionCallBack), (void*)0xbebabeba);
			}
			else
			{
				callBack = CCCallFuncND::create(this, callfuncND_selector(GameScenePlayLayer::runJumpActionCallBack), (void*)0xbebabebb);
			}
				
            CCFiniteTimeAction*  action = CCSequence::create(jumpAction,callBack,NULL);
            imManArmature->runAction(action);
        }
        
        if(0 == armatureName.compare("IMRunStop"))
        {
            this ->IMStandJump();
            CCActionInterval * jumpAction = CCJumpTo::create(0.5,CCPointMake(imManArmature->getPosition().x,imManArmature->getPosition().y),200,1);
            CCCallFunc * callBack = CCCallFuncND::create(this, callfuncND_selector(GameScenePlayLayer::standJumpActionCallBack), (void*)0xbebabeba);
            CCFiniteTimeAction*  action = CCSequence::create(jumpAction,callBack,NULL);
            imManArmature->runAction(action);
        }
	}
    
	if(nMoveY<-10 && fabs(tan(nMoveY/nMoveX))>fabs(sqrt(3)/radian))
	{
		;
	}
}

void GameScenePlayLayer::IMRunning()
{
	CCArmature *armature = NULL;
	armature = CCArmature::create("IMRun");
	armature->getAnimation()->play("Running");
	armature->getAnimation()->setSpeedScale(2.0f);
	armature->setScale(playerScale);
	armature->setAnchorPoint(ccp(0.5,0));
	armature->setPosition(ccp(playerX+30, playerY));
	amaturePosition = armature->getPosition();
	addChild(armature);
	imManArmature = armature;
	actionNum = ACTION_RUN;
	GameScene::shareGameScene()->gameSceneMapLayer->move();
}

void GameScenePlayLayer::IMStandJump()
{
	CCArmature *armature = NULL;
	armature = CCArmature::create("IMStandJump");
	armature->getAnimation()->play("StandJump");
	armature->getAnimation()->setSpeedScale(1.5f);
	armature->setScale(playerScale);
	armature->setAnchorPoint(ccp(0.5,0));
	armature->setPosition(ccp(playerX+20, playerY));
	amaturePosition = armature->getPosition();
	addChild(armature);
	imManArmature = armature;
	actionNum = ACTION_STAND_JUMP;
	GameScene::shareGameScene()->gameSceneMapLayer->stop();
}

void GameScenePlayLayer::IMRunJump()
{  
 	CCArmature *armature = NULL;
	armature = CCArmature::create("IMRunJump");
	armature->getAnimation()->play("RuningJump");
	armature->getAnimation()->setSpeedScale(1.5f);
	armature->setScale(playerScale);
	armature->setAnchorPoint(ccp(0.5,0));
	armature->setPosition(ccp(playerX+20, playerY));
	amaturePosition = armature->getPosition();
	addChild(armature);
	imManArmature = armature;
	actionNum = ACTION_RUN_JUMP;
	GameScene::shareGameScene()->gameSceneMapLayer->move();
	GameScene::shareGameScene()->gameSceneMapLayer->setMovedSpeed(6);
	armature->getAnimation()->setMovementEventCallFunc(this, movementEvent_selector(GameScenePlayLayer::amatureActionCallBack));
}

void GameScenePlayLayer::IMRunningStop()
{
	CCArmature *armature = NULL;
	armature = CCArmature::create("IMRunStop");
	armature->getAnimation()->play("RunningStop");
	armature->getAnimation()->setSpeedScale(1.5f);
	armature->setScale(playerScale);
	armature->setAnchorPoint(ccp(0.5,0));
	armature->setPosition(ccp(playerX+50, playerY));
	amaturePosition = armature->getPosition();
	addChild(armature);
	imManArmature = armature;
	actionNum = ACTION_RUN_STOP;
	GameScene::shareGameScene()->gameSceneMapLayer->stop();
}

void GameScenePlayLayer::IMRunAttack(CCPoint touch)
{
    float angle = getAngle(touch);
    CCPoint posHand = getPosHand(angle);
    
    CCArmature *armature = NULL;
    armature = CCArmature::create("LaserRunAttack");
    armature->getAnimation()->play("RunningAttack");
    armature->getAnimation()->setSpeedScale(2.0);
	armature->setAnchorPoint(ccp(0.5,0));
	armature->setScale(playerScale);
	armature->setPosition(ccp(playerX+40, playerY));
    //CCBone* leftArmBone = armature->getBone("LeftTopArmAttack");
    //leftArmBone->setRotation(getAngle(touch));
    amaturePosition = armature->getPosition();
	addChild(armature);
	imManArmature = armature;
    armature->getAnimation()->setMovementEventCallFunc(this, movementEvent_selector(GameScenePlayLayer::setAttackEvent));
    
    _attackPos = posHand;
    _attackDir = angle;
}

void GameScenePlayLayer::IMStandAttack(CCPoint touch)
{
    CCPoint posHand = getPosHand(0.1);
    _attackPos = posHand;
    float angle = getAngle(touch);
    
    CCArmature *armature = NULL;
    armature = CCArmature::create("LaserStandAttack");
    armature->getAnimation()->play("StandAttack");
    armature->getAnimation()->setSpeedScale(0.5);
	armature->setAnchorPoint(ccp(0.5,0));
	armature->setScale(playerScale);
	armature->setPosition(ccp(playerX, playerY));
    addChild(armature);
    imManArmature = armature;
    armature->getAnimation()->setMovementEventCallFunc(this, movementEvent_selector(GameScenePlayLayer::setAttackEvent));
	
    _attackDir = angle;
}
void GameScenePlayLayer::IMDeath()
{
	this->setTouchEnabled(false);
	imManArmature->removeFromParentAndCleanup(true);
	CCArmature *armature = NULL;
	armature = CCArmature::create("IMDead");
	armature->getAnimation()->playByIndex(0.0f, 1.0f, 1.0f,0.0f, 1.0f);
	armature->getAnimation()->setSpeedScale(1.0f);
	armature->setScale(playerScale);
	armature->setAnchorPoint(ccp(0.5,0));
	armature->setPosition(ccp(100, playerY));
	amaturePosition = armature->getPosition();
	addChild(armature);
	imManArmature = armature;
	actionNum = ACTION_DEATH;
	armature->getAnimation()->setMovementEventCallFunc(this, movementEvent_selector(GameScenePlayLayer::Dead));
}

float GameScenePlayLayer::getAngle(CCPoint touch)
{
    //touch = ccp(240, 290);
    //CCPoint posOrg = ccp(108, 105);
    CCPoint posOrg = ccp(135, 132);
    if(touch.x <= posOrg.x)
        return -1.57;   //up max->90 degreeã€?
    if (touch.y == posOrg.y) {
        if (touch.x > posOrg.x) {
            return 0;
        }
    }
    
    float tan = (touch.y - posOrg.y)/(touch.x - posOrg.x);
    if (tan < -6) {
        tan = -6;    //down max->45 degreeã€?
    }
    double angle = atan(tan);
    return -angle;
}
CCPoint GameScenePlayLayer::getPosHand(float angle)
{
    CCPoint posH = ccp(141, 141);
    
    return posH;
}
void GameScenePlayLayer::setAttackEvent(cocos2d::extension::CCArmature *armature, MovementEventType movementType, const char *movementID)
{
    std::string id = movementID;
    
    if (movementType == COMPLETE || movementType == LOOP_COMPLETE)
    {
        AudioPlayer::sharedAudio()->playEffect(Effect_Attack_0);
        GameScene::shareGameScene()->laser->addLaser(_attackPos, _attackDir);
        isAttack = false;
		imManArmature->stopAllActions();
		imManArmature->removeFromParentAndCleanup(false);		
		if(actionNum == ACTION_RUN)
		{
			this ->IMRunning();
		}
		else if(actionNum == ACTION_RUN_STOP)
		{
			this ->IMRunningStop();
		}
    }
}
void GameScenePlayLayer::amatureActionCallBack(cocos2d::extension::CCArmature *armature, MovementEventType movementType, const char *movementID)
{
	std::string id = movementID;
	if (movementType == COMPLETE || movementType == LOOP_COMPLETE)
	{
		switch (actionNum)
		{
		case ACTION_RUN_JUMP:
		{
			 GameScene::shareGameScene()->gameSceneMapLayer->setMovedSpeed(3);
		}
			 break;
		default:
			break;
		}
	}
}
bool GameScenePlayLayer::checkIfTouchNotInSetBtnArea(CCPoint touchPosition, CCSize setBtnSize, CCPoint setBtnPosition)
{
	if(touchPosition.x < setBtnPosition.x-(setBtnSize.width/2)  ||
	   touchPosition.x > setBtnPosition.x+(setBtnSize.width/2)  ||
	   touchPosition.y < setBtnPosition.y-(setBtnSize.height/2) ||
	   touchPosition.y > setBtnPosition.y-(setBtnSize.height/2)){
		return true;
	}else{
		return false;
	}
}

void GameScenePlayLayer::Dead(cocos2d::extension::CCArmature *armature, MovementEventType movementType, const char *movementID)
{
 	GameScene::shareGameScene()->gameOver();
}

int GameScenePlayLayer::getMonsterGroundAmount()
{
	
	return monsterGroundAmount;
}

int GameScenePlayLayer::getMonsterSkyAmount()
{
	return monsterSkyAmount;
}
 void GameScenePlayLayer::changeSpeed(float t)
 {
	imManArmature->getAnimation()->setSpeedScale(2.0f);
	GameScene::shareGameScene()->gameSceneMapLayer->setMovedSpeed(3);
 }

File: cocos2d-html5_i/IronCity/src/layer/GameScenePlayLayer.js

var GameScenePlayLayer = cc.Layer.extend
({
    isMouseDown:false,
    actionNum:null,
    imManArmature:null,
    armaturePosition:null,
    isAttack:false,
    _attackPos:null,
    _attackDir:null,

    monsterGroundAmount:0,
    monsterSkyAmount:0,
    touchTime:0,
    imManArmatureBrood:0,
    m_tBeginPos:null,
    s_tCurPos:null,
    playerBoundingBox:null,
    playerScale:0,
    playerX:0,
    playerY:0,

    ACTION_CROUCH:0,
    ACTION_RUN:1,
    ACTION_STAND_JUMP:2,
    ACTION_RUN_JUMP:3,
    ACTION_CROUCH_JUMP:4,
    ACTION_RUN_STOP:5,
    ACTION_DEATH:6,

    //init function.
    init:function () {
        this._super();
        this.setTouchEnabled(true);
        //this.setTouchMode(cc.TOUCH_ONE_BY_ONE);
        var size = cc.Director.getInstance().getWinSize();

        //add cocostudio json file to widget.
        cc.ArmatureDataManager.getInstance().addArmatureFileInfo(Json_CMRun);
        cc.ArmatureDataManager.getInstance().addArmatureFileInfo(Json_CMRunJump);
        cc.ArmatureDataManager.getInstance().addArmatureFileInfo(Json_CMRunStop);
        cc.ArmatureDataManager.getInstance().addArmatureFileInfo(Json_CMStandJump);
        cc.ArmatureDataManager.getInstance().addArmatureFileInfo(Json_LaserRunAttack);
        cc.ArmatureDataManager.getInstance().addArmatureFileInfo(Json_LaserStandAttack);
        cc.ArmatureDataManager.getInstance().addArmatureFileInfo(Json_CMDead);

        this.playerX = 50.0;
        this.playerY = 70.0;
        this.playerScale = 0.6;
        this.imManArmatureBrood = 100;

        this.monsterGroundAmount = 0;
        this.monsterSkyAmount = 0;

        this.touchTime = 0;
        this.isAttack = false;
        this._attackPos = cc.p(0,0);
        this._attackDir = 0.0;
        this.CMRunningStop();
        this.actionNum = this.ACTION_RUN_STOP;
    },
    //callback function of runjump action.
    runJumpActionCallBack:function(sender, data){
        this.imManArmature.stopAllActions();
        this.imManArmature.removeFromParent(false);
        if(0xbebabebb == data)
        {
            this.CMRunning();
        }
        else
        {
            this.CMRunningStop();
        }
    },
    //callback function of standjump action.
    standJumpActionCallBack:function(){
        this.imManArmature.stopAllActions();
        this.imManArmature.removeFromParent(false);
        this.CMRunningStop();
    },
    //register touch dispatcher.
    registerWithTouchDispatcher:function(){
        cc.registerTargetedDelegate(0, false, this);
    },
    //callback function of menu close button.
    menuCloseCallback:function(){
        var splitCols = cc.MoveTo.create(1.0 ,cc.p(this.imManArmature.getPosition().x+300, this.imManArmature.getPosition().y));
        this.imManArmature.runAction(splitCols);
    },
    //change IronMan's speed.
    changeSpeed:function(t){
        this.imManArmature.getAnimation().setSpeedScale(2.0);
        GameScene.getScene().moveMap.setMovedSpeed(3);
    },
    //on touch began.
    onTouchBegan:function (touches, event) {
        this.m_tBeginPos = touches.getLocation();

        return true;
    },
    //on touch moved.
    onTouchMoved:function (touches, event) {
        this.touchTime++;
    },
    //on touch ended.
    onTouchEnded:function (touches, event) {
        if(this.isAttack){
            return;
        }
        if(this.touchTime>30)
        {
            this.touchTime = 0;
            return;
        }

        var touchLocation  = touches.getLocation();
        var setBtnLocation = GameScene.getScene().menuLayer.settingBtn.getPosition();
        var setBtnSize = GameScene.getScene().menuLayer.settingBtn.getSize();

        if(this.touchTime<6 && this.checkIfTouchNotInSetBtnArea(touchLocation,setBtnSize, setBtnLocation))
        {
            if(this.actionNum != this.ACTION_RUN && this.actionNum != this.ACTION_RUN_STOP)
            {
                return;
            }
            this.isAttack = true;
            this.imManArmature.stopAllActions();
            this.imManArmature.removeFromParent(false);
            if(this.actionNum == this.ACTION_RUN)
            {
                this.CMRunAttack(touchLocation);
            }
            else if(this.actionNum == this.ACTION_RUN_STOP)
            {
                this.CMStandAttack(touchLocation);
            }

            this.touchTime = 0;
            return;
        }

        var touchLocation = touches.getLocation();
        var nMoveX = touchLocation.x - this.m_tBeginPos.x;
        var nMoveY = touchLocation.y - this.m_tBeginPos.y;
        var radian = 10;

        if(nMoveX>10 && Math.abs(Math.tan(nMoveY/nMoveX))<Math.abs(Math.sqrt(3)/radian))
        {
            if(this.actionNum == this.ACTION_RUN)
            {
                if(GameScene.getScene().moveMap.getMoveSpeed()!=6)
                {
                    GameScene.getScene().moveMap.setMovedSpeed(6);
                    this.imManArmature.getAnimation().setSpeedScale(3.0);
                    this.schedule(this.changeSpeed,0.5, 0, 0.0);
                    return;
                }
            }
            this.imManArmature.stopAllActions();
            this.imManArmature.removeFromParent(false);
            this.CMRunning();
        }

        if(nMoveX<-10 && Math.abs(Math.tan(nMoveY/nMoveX))<Math.abs(Math.sqrt(3)/radian))
        {
            if(this.actionNum == this.ACTION_RUN_STOP)
                return;

            this.imManArmature.stopAllActions();
            this.imManArmature.removeFromParent(false);
            this.CMRunningStop();
        }

        this.touchTime = 0;

        if(nMoveY>10 && Math.abs(Math.tan(nMoveY/nMoveX))>Math.abs(Math.sqrt(3)/radian))
        {
            if(this.actionNum == this.ACTION_STAND_JUMP || this.actionNum == this.ACTION_RUN_JUMP)
                return;

            var armatureName = this.imManArmature.getName();
            this.imManArmature.stopAllActions();
            this.imManArmature.removeFromParent(false);

            if(armatureName == "CMRun")
            {
                this.CMRunJump();
                var jumpAction = cc.JumpTo.create(0.5,cc.p(this.imManArmature.getPosition().x,this.imManArmature.getPosition().y),100,1);
                var callBack;
                if(nMoveX<0)
                {
                    callBack = cc.CallFunc.create(this.runJumpActionCallBack, this, 0xbebabeba);
                }
                else
                {
                    callBack = cc.CallFunc.create(this.runJumpActionCallBack, this, 0xbebabebb);
                }
                var action = cc.Sequence.create(jumpAction,callBack);
                this.imManArmature.runAction(action);
            }

            if(armatureName == "CMRunStop")
            {
                this.CMStandJump();
                var jumpAction = cc.JumpTo.create(0.5,cc.p(this.imManArmature.getPositionX(),this.imManArmature.getPositionY()),200,1);
                var callBack = cc.CallFunc.create(this.standJumpActionCallBack, this, 0xbebabeba);
                var action = cc.Sequence.create(jumpAction,callBack);
                this.imManArmature.runAction(action);
            }
        }
    },
    //on touch cancelled.
//    onTouchesCancelled:function (touches, event) {
//    },


    //Running animation.
    CMRunning:function(){
        var armature = cc.Armature.create("CMRun");
        armature.getAnimation().play("Running");
        armature.getAnimation().setSpeedScale(2.0);
        armature.setScale(this.playerScale);
        armature.setAnchorPoint(cc.p(0.5,0));
        armature.setPosition(cc.p(this.playerX+30, this.playerY));
        this.armaturePosition = armature.getPosition();
        this.addChild(armature);
        this.imManArmature = armature;

        this.actionNum = this.ACTION_RUN;
        if(GameScene.getScene() && GameScene.getScene().moveMap)
            GameScene.getScene().moveMap.move();
    },

    //Run jump animation.
    CMRunJump:function(){
        var armature = cc.Armature.create("CMRunJump");
        armature.getAnimation().play("RuningJump");
        armature.getAnimation().setSpeedScale(1.5);
        armature.setScale(this.playerScale);
        armature.setAnchorPoint(cc.p(0.5,0));
        armature.setPosition(cc.p(this.playerX+20, this.playerY));
        this.armaturePosition = armature.getPosition();
        this.addChild(armature);
        this.imManArmature = armature;
        this.actionNum = this.ACTION_RUN_JUMP;
        GameScene.getScene().moveMap.setMovedSpeed(6);
        armature.getAnimation().setMovementEventCallFunc(this.amatureActionCallBack, this);
    },

    //Stand jump animation.
    CMStandJump:function(){
        var armature = cc.Armature.create("CMStandJump");
        armature.getAnimation().play("StandJump");
        armature.getAnimation().setSpeedScale(1.5);
        armature.setScale(this.playerScale);
        armature.setAnchorPoint(cc.p(0.5,0));
        armature.setPosition(cc.p(this.playerX+20, this.playerY));
        this.armaturePosition = armature.getPosition();
        this.addChild(armature);
        this.imManArmature = armature;
        this.actionNum = this.ACTION_STAND_JUMP;
        if(GameScene.getScene() && GameScene.getScene().moveMap)
            GameScene.getScene().moveMap.stop();
    },

    //Running stop animation.
    CMRunningStop:function(){
        var armature = cc.Armature.create("CMRunStop");
        armature.getAnimation().play("RunningStop");
        armature.getAnimation().setSpeedScale(1.5);
        armature.setScale(this.playerScale);
        armature.setAnchorPoint(cc.p(0.5,0));
        armature.setPosition(cc.p(this.playerX+50, this.playerY));
        this.armaturePosition = armature.getPosition();
        this.addChild(armature);
        this.imManArmature = armature;

        this.actionNum = this.ACTION_RUN_STOP;
        if(GameScene.getScene() && GameScene.getScene().moveMap)
            GameScene.getScene().moveMap.stop();
    },
    //Running attack animation.
    CMRunAttack:function(touch){
        var angle = this.getAngle(touch);
        var posHand = this.getPosHand(angle);

        var armature = cc.Armature.create("LaserRunAttack");
        armature.getAnimation().play("RunningAttack");
        armature.getAnimation().setSpeedScale(2.0);
        armature.setScale(this.playerScale);
        armature.setAnchorPoint(cc.p(0.5,0));
        armature.setPosition(cc.p(this.playerX+40, this.playerY));
        this.armaturePosition = armature.getPosition();
        this.addChild(armature);
        this.imManArmature = armature;
        armature.getAnimation().setMovementEventCallFunc(this.setAttackEvent, this);

        this._attackPos = posHand;
        this._attackDir = angle;
    },
    //Stand attack animation.
    CMStandAttack:function(touch){
        var posHand = this.getPosHand(0.1);
        this._attackPos = posHand;
        var angle = this.getAngle(touch);

        var armature = cc.Armature.create("LaserStandAttack");
        armature.getAnimation().play("StandAttack");
        armature.getAnimation().setSpeedScale(1.5);
        armature.setScale(this.playerScale);
        armature.setAnchorPoint(cc.p(0.5,0));
        armature.setPosition(cc.p(this.playerX, this.playerY));
        this.armaturePosition = armature.getPosition();
        this.addChild(armature);
        this.imManArmature = armature;
        armature.getAnimation().setMovementEventCallFunc(this.setAttackEvent, this);

        this._attackDir = angle;
    },
    //death animation.
    CMDeath:function(){
        this.setTouchEnabled(false);
        this.imManArmature.removeFromParent(true);
        var armature = null;
        armature = cc.Armature.create("CMDead");
        armature.getAnimation().playByIndex(0.0, 1.0, 1.0,0.0, 1.0);
        armature.getAnimation().setSpeedScale(0.5);
        armature.setScale(this.playerScale);
        armature.setAnchorPoint(cc.p(0.5,0));
        armature.setPosition(cc.p(100, this.playerY-8));
        this.armaturePosition = armature.getPosition();
        this.addChild(armature);
        this.imManArmature = armature;
        this.actionNum = this.ACTION_DEATH;
        armature.getAnimation().setMovementEventCallFunc(this.Dead, this);
    },
    //CallBack of running jump animation.
    amatureActionCallBack:function(armature, movementType, armature){
        if (movementType == CC_MovementEventType_COMPLETE || movementType == CC_MovementEventType_LOOP_COMPLETE)
        {
            switch (this.actionNum)
            {
                case this.ACTION_RUN_JUMP:
                {
                    GameScene.getScene().moveMap.setMovedSpeed(3);
                }
                    break;
                default:
                    break;
            }
        }
    },
    //get an angle
    getAngle:function(touch){
        var posOrg = cc.p(135, 132);
        if(touch.x <= posOrg.x)
            return -1.57;   //up max->90 degree„Ä?
        if (touch.y == posOrg.y) {
            if (touch.x > posOrg.x) {
                return 0;
            }
        }

        var tan = (touch.y - posOrg.y)/(touch.x - posOrg.x);
        if (tan < -6) {
            tan = -6;    //down max->45 degree„Ä?
        }
        var angle = Math.atan(tan);
        return -angle;
    },
    //get IronMan's hand position.
    getPosHand:function(angle){
        var posH = cc.p(141, 141);
        return posH;
    },
    //callback function of attack animation.
    setAttackEvent:function(armature, movementType, movementID){
        if (movementType == CC_MovementEventType_COMPLETE || movementType == CC_MovementEventType_LOOP_COMPLETE)
        {
            //play audio and launch laser.
            AudioPlayer.getInstance().playEffect(g_ArrEffects.Effect_Attack_0);
            GameScene.getScene().laser.addLaser(this._attackPos, this._attackDir);
            this.isAttack = false;
            this.imManArmature.stopAllActions();
            this.imManArmature.removeFromParent(false);
            if(this.actionNum == this.ACTION_RUN)
            {
                this.CMRunning();
            }
            else if(this.actionNum == this.ACTION_RUN_STOP)
            {
                this.CMRunningStop();
            }
        }
    },
    //callback function of dead animation.
    Dead:function(armature, movementType, movementID){
        if (movementType == CC_MovementEventType_COMPLETE || movementType == CC_MovementEventType_LOOP_COMPLETE)
        {
            if(GameScene.getScene())
                GameScene.getScene().gameOver();
        }
    },
    //chech area of not in set button.
    checkIfTouchNotInSetBtnArea:function(touchPosition, setBtnSize, setBtnPosition){
        return true;
    },
    //get Amount of monster on ground.
    getMonsterGroundAmount:function(){
        return this.monsterGroundAmount;
    },
    //get Amount of monster in sky.
    getMonsterSkyAmount:function(){
        return this.monsterSkyAmount;
    },
    //add Amount of monster on ground.
    addMonsterGroundAmount:function(){
        this.monsterGroundAmount ++;
    },
    //add Amount of monster in sky.
    addMonsterSkyAmount:function(){
        this.monsterSkyAmount ++;
    },
    pause:function()
    {
        this.imManArmature.pauseSchedulerAndActions();
        this.imManArmature.unscheduleUpdate();
    },
    play:function()
    {
        this.imManArmature.resumeSchedulerAndActions();
        this.imManArmature.scheduleUpdate();
    }

});
